aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1908,7 +1911,84 @@ Tensor& vdot_out(const Tensor& self, const Tensor& other, Tensor& result) {
   return result.fill_(self.vdot(other));
 }
 
-static bool should_fold(const Tensor& tensor1, const Tensor& tensor2) {
+#define AUTOTUNE_MAX_DIM 4
+
+struct MatmulAutotuneCacheKey {
+  int tensor1_dim[AUTOTUNE_MAX_DIM];
+  int tensor1_stride[AUTOTUNE_MAX_DIM];
+  int tensor2_dim[AUTOTUNE_MAX_DIM];
+  int tensor2_stride[AUTOTUNE_MAX_DIM];
+  bool should_fold;
+  ScalarType dtype;
+  bool is_cuda;
+  int device;
+};
+
+struct MatmulAutotuneCacheKeyWrapper : ParamsWrapper<MatmulAutotuneCacheKey> {
+  MatmulAutotuneCacheKeyWrapper() = default;
+
+  MatmulAutotuneCacheKeyWrapper(
+    const Tensor& tensor1,
+    const Tensor& tensor2,
+    bool should_fold) {
+    TORCH_INTERNAL_ASSERT(tensor1.sizes().size() <= AUTOTUNE_MAX_DIM);
+    TORCH_INTERNAL_ASSERT(tensor2.sizes().size() <= AUTOTUNE_MAX_DIM);
+    std::copy(tensor1.sizes().begin(), tensor1.sizes().end(), this->pod.tensor1_dim);
+    std::copy(tensor1.strides().begin(), tensor1.strides().end(), this->pod.tensor1_stride);
+    std::copy(tensor2.sizes().begin(), tensor2.sizes().end(), this->pod.tensor2_dim);
+    std::copy(tensor2.strides().begin(), tensor2.strides().end(), this->pod.tensor2_stride);
+    this->pod.should_fold = should_fold;
+    this->pod.is_cuda = tensor1.device().type() == at::kCUDA;
+    this->pod.dtype = tensor1.scalar_type();
+    this->pod.device = at::detail::getCUDAHooks().current_device();
+  }
+};
+
+template <typename T, typename KeyType>
+struct MatmulAutotuneCache {
+std::unordered_map<KeyType, T, ParamsWrapperHash<KeyType>> engine_cache;
+
+T* find(const KeyType& key) {
+  auto it = engine_cache.find(key);
+  if (it == engine_cache.end()) {
+    return nullptr;
+  }
+  return &(it->second);
+}
+
+void update(const KeyType& key, T& results) {
+  engine_cache.erase(key);
+  engine_cache.emplace(key, std::move(results));
+}
+
+};
+
+MatmulAutotuneCache<float, MatmulAutotuneCacheKeyWrapper> matmul_autotune_cache;
+
+static bool should_fold(const Tensor& tensor1, const Tensor& tensor2, MatmulAutotuneCacheKeyWrapper & key, bool& need_profiling, const bool autotune) {
+  if (autotune) {
```

Comment:
This should be moved just before the line
```
  // t1->view(-1, t1->size(-1)) does not copy only when the first n-1 dimensions are contiguous
```
In all the other cases we are quite sure we want to fold.

Reply:
I did a bit of twiddling and found out what makes the biggest difference is moving the autotuning before 
```
2022   bool t2_requires_grad = tensor1_larger ? tensor2.requires_grad() : tensor1.requires_grad();
2023   if (t2_requires_grad && !has_out) {
```

Reply:
That doesn't make much sense as, in your tests, the tensors should not have requires_grad or out, no?

Reply:
`tensor1` does have `requires_grad=True` in the test but has smaller `dim`, so I'm assuming some speedup comes from autotuning not folding when the heuristic would fold?

Reply:
If it has `requires_grad=True` you should benchmark forward + backward to make it realistic. Not folding in those cases would incur in a large tensor being materialised in the backward.